Correctly dump check constraints for MySQL 8.0.16+

stevehill1981 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created because Rails doesn't currently handle check constraints correctly in newer versions of MySQL 8.0.

Fixes #47849.

### Detail

If you're using MySQL 8.0.16+ and your database contains a table with a check constraint, the first and last characters of the constraint will be stripped when dumping the schema; this makes it impossible to use check constraints in a MySQL 8.0 database with the `:ruby` schema format; once dumped, they cannot be re-imported.

This is because, up until MySQL 8.0.16, all check constraints were surrounded by an extra set of parentheses; this behaviour differed from that presented by MariaDB, which handled them correctly thanks to an engine check within the AbstractMySqlAdapter.

With this change, we continue to handle MariaDB correctly, but also add an extra database version check so that MySQL 8.0.16+ behaves in the same way.

It is also possible that the exported check constraint will contain invalid and unexpected whitespace characters; we handle that here too, stripping out sequences of `\\` and `\n`, as well as any other whitespace.

### Additional information

I've tested this as thoroughly as possible against local versions of MySQL 8.0 running in Docker. I wasn't able to figure out how to write a test for the new behaviour; there's only a single test that deals with check constraints in the schema dumper and it doesn't have any MySQL-specific behaviour.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

stevehill1981 committed:
Correctly dump check constraints for MySQL 8.0.16+

If you're using MySQL 8.0.16+ and your database contains a table with
a check constraint, the first and last characters of the constraint
will be stripped when dumping the schema; this makes it impossible
to use check constraints in a MySQL 8.0 database with the `:ruby`
schema format; once dumped, they cannot be re-imported.

This is because, up until MySQL 8.0.16, all check constraints were
surrounded by an extra set of parentheses; this behaviour differed
from that presented by MariaDB, which handled them correctly thanks
to an engine check within the AbstractMySqlAdapter.

With this change, we only strip the parentheses if they are present,
which prevents the corruption of the constraint.

It is also possible that the exported check constraint will contain
invalid and unexpected whitespace characters; we handle that here too,
stripping out sequences of `\\` and `\n`, as well as any other
whitespace.

stevehill1981 commented:
I've arrived at a somewhat different solution here than I originally set out with; the new code now only strips the parentheses if they're present, which fixes the original bug where the constraint was being corrupted as a result of removing the first and last characters indiscriminately.

stevehill1981 commented:
I've figured out where to add a test for this, so I'll update with that tomorrow.

byroot reviewed (commented):
Please squash you commits.

stevehill1981 commented:
@byroot Thanks! Squashed the commits, added the test I mentioned (and fixed the others I broke along the way). There are still some test failures in Buildkite, but of the 5 failures only one of them is related to my change - and it weirdly only fails on the Ruby 3.2 tests, all others are fine.

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -648,6 +649,21 @@ def check_version # :nodoc:
+        expression = expression.gsub('\\', "")
+        expression = expression.gsub(/\s{2,}/, " ")
+        # rubocop:enable Style/StringLiterals
+        # rubocop:enable Performance/StringReplacement
```
We don't generally do rubocop enable/disable in the files because it pollutes them. We instead add them to the exclude list. 

This method should also be moved to the private namespace, we don't want applications to call this method.

stevehill1981 replied:
Will tweak that this afternoon, thanks @eileencodes.

stevehill1981 replied:
Tweaks made!

stevehill1981 replied:
I had to add a couple of extra files to the exclusions for `Performance/StringReplacement`, when I excluded the file I edited there were other changes that Rubocop flagged. Let me know if those need to be removed again.

byroot commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -714,6 +715,16 @@ def extract_precision(sql_type)
+          # These statements use single quotes because the expression may contain
+          # strings of these characters, _not_ the escaped versions that using
+          # double quotes would produce.
+          expression = expression.gsub('\n', "")
+          expression = expression.gsub('\\', "")
```
It's unclear to me if you are trying to strip `\` or `\\`.

Also it would make sense to merge these two gsub together as a regexp, that would also fix the rubocop issue.

stevehill1981 replied:
👍🏻 I'll tweak that next.

stevehill1981 replied:
Done 😁 Thanks for the detailed feedback on this.

stevehill1981 commented on .rubocop.yml:
```diff
@@ -319,6 +321,10 @@ Performance/ReverseEach:
+  Exclude:
+    - 'activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb'
+    - 'activerecord/test/cases/validations_test.rb'
+    - 'activesupport/test/xml_mini_test.rb'
```
This is what came out of Rubocop when I ran it:

```
activesupport/test/xml_mini_test.rb:356:36: C: [Correctable] Performance/StringReplacement: Use tr instead of gsub.
      assert_equal expected_base64.gsub(/\n/, " ").strip, parser.call(base64)
                                   ^^^^^^^^^^^^^^^
activesupport/test/xml_mini_test.rb:360:36: C: [Correctable] Performance/StringReplacement: Use tr instead of gsub.
      assert_equal expected_base64.gsub(/\n/, " ").strip, parser.call(base64, "encoding" => "base64")
                                   ^^^^^^^^^^^^^^^
```

byroot replied:
Not sure what is going on, but you haven't touched that file, and it passes CI. Anyway, I don't think you need any of these exclusion, the code can both be made more performant and not need any exception.

stevehill1981 replied:
Thanks, @byroot - I've removed the exclusions and merged the `gsub` calls as you suggested above.

stevehill1981 replied:
I've just added `if_exists: true` to the `remove_check_constraint` call, I'm assuming it's just a timing issue but hopefully this will help me figure out what's going wrong otherwise.

stevehill1981 commented on .rubocop.yml:
```diff
@@ -197,6 +197,8 @@ Layout/SpaceInsideParens:
 Style/StringLiterals:
   Enabled: true
   EnforcedStyle: double_quotes
+  Exclude:
+    - 'activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb'
```
Just spotted this! Pushing a new version now.

eileencodes commented on activerecord/test/cases/migration/check_constraint_test.rb:
```diff
@@ -44,6 +54,18 @@ def test_check_constraints
             assert_equal "price > discounted_price", constraint.expression
           end
 
+          if current_adapter?(:Mysql2Adapter)
```
Code looks good but this new test is flaky https://buildkite.com/rails/rails/builds/95406#01874c8b-219b-4629-8c29-d7070ffbf1cf/1066-1075 (didn't fail on all mysql builds but did fail on some).

stevehill1981 replied:
With my last change, it succeeds everywhere EXCEPT for MariaDB... which looks like it might be because the `JSON_SCHEMA_VALID` function only exists in MySQL 8.0. I'm not sure if we need to care about that or not; I'm inclined to switch to another function that definitely does exist.

stevehill1981 replied:
I've moved to `JSON_CONTAINS`, which looks like it's fine on both MySQL and MariaDB. It wasn't important in this case which function call was being used, just that it was correctly handled when it _was_ used.

eileencodes commented:
Thanks @stevehill1981! Congrats on your first Rails commit 🎉

stevehill1981 commented:
Thank you @eileencodes and @byroot for shepherding me through this process and giving such useful feedback! I'm really pleased to see this landing 😃

chen-anders commented:
We just got hit by this issue after we switched t usingo MySQL 8.0.32 - I see that this change did not make it to the respective 7.0.6 release - would there be a timeline for that?

byroot commented:
We didn't think of backporting it. 

I just did at 0efee08e9fd7f97f3ddfeef64b821db129544d3a, you can point your gemfile at the 7-0-stable branch.
